Aptoide/aptoide-client-v8
# [task/ASV-1800] Refactor AppView load app

**What does this PR do?**

   This PR refactors the initial load of the app view on the presenter side. As a side effect, many other things that are listed below were refactored. This PR came because of visual issues with the new DownloadProgressView that occured as a result of the mess that was loadApp().

List of refactored things:
1. Major changes to AppViewModel.
   1. Renamed old AppViewModel to AppModel. This contains the mapping from DetailedApp.
   2. New AppViewModel class. This class contains other Models that are essential to the AppView and used in different components. This includes AppModel, DownloadModel, AppCoinsViewModel and MigrationModel. This also makes it easy to add new models if need be.
   3. New AppViewModelManager class. The new AppViewModel is managed by this new Manager, which is responsible for caching relevant data and efficiently providing AppViewModel.
2. Major flow changes to the loadApp() chain.
   1. Un-nested and separated each action to its own type. What this means is that every view operation is together, every analytic operation is together, etc. This makes it very easy to follow what's happening and where to add relevant code.
   2. Extracted methods to self-document code.
   3. Refactored AppView open types to avoid duplicated code. Fixed bug with OPEN_AND_INSTALL where the action was hardcoded.

Among other minor things.

**Database changed?**

   No

**Where should the reviewer start?**

- [ ] AppViewModel.java

**How should this be manually tested?**

  There's a lot of changes to the appview presenter/manager code, so it needs to be tested thoroughly. Just try to break the appview in general, but the main focus is the initialization. Test downloads, check appc apps, etc. Don't forget to test Appc Promotions because there were some changes there as well (check https://github.com/Aptoide/aptoide-client-v8/pull/1016  to see how to test).

**What are the relevant tickets?**

  [ASV-1800](https://aptoide.atlassian.net/browse/ASV-1800)


**Code Review Checklist**
